libflux/src/semantic/analyze.rs

```diff
@@ -0,0 +1,2604 @@
+        Ok(Package {
+            loc: &pkg.base.location,
+            package: pkg.package.clone(),
+            files: files,
+        })
+    }
+
+    fn analyze_file<'a>(&mut self, file: &'a ast::File) -> Result<File<'a>> {
+        let package = self.analyze_package_clause(&file.package)?;
+        let imports = file
+            .imports
+            .iter()
+            .map(|i| self.analyze_import_declaration(i))
+            .collect::<Result<Vec<ImportDeclaration>>>()?;
+        let body = file
+            .body
+            .iter()
+            .map(|s| self.analyze_statement(s))
+            .collect::<Result<Vec<Statement>>>()?;
+        Ok(File {
+            loc: &file.base.location,
+            package,
+            imports,
+            body,
+        })
+    }
+
+    fn analyze_package_clause<'a>(
+        &mut self,
+        pkg: &'a Option<ast::PackageClause>,
+    ) -> Result<Option<PackageClause<'a>>> {
+        if pkg.is_none() {
+            return Ok(None);
+        }
+        let pkg = pkg.as_ref().unwrap();
+        let name = self.analyze_identifier(&pkg.name)?;
+        Ok(Some(PackageClause {
+            loc: &pkg.base.location,
+            name,
+        }))
+    }
+
+    fn analyze_import_declaration<'a>(
+        &mut self,
+        imp: &'a ast::ImportDeclaration,
+    ) -> Result<ImportDeclaration<'a>> {
+        let alias = match &imp.alias {
+            None => None,
+            Some(id) => Some(self.analyze_identifier(&id)?),
+        };
+        let path = self.analyze_string_literal(&imp.path)?;
+        Ok(ImportDeclaration {
+            loc: &imp.base.location,
+            alias,
+            path,
+        })
+    }
+
+    fn analyze_statement<'a>(&mut self, stmt: &'a ast::Statement) -> Result<Statement<'a>> {
+        match &stmt {
+            ast::Statement::Option(s) => Ok(Statement::Option(self.analyze_option_statement(&s)?)),
+            ast::Statement::Builtin(s) => {
+                Ok(Statement::Builtin(self.analyze_builtin_statement(&s)?))
+            }
+            ast::Statement::Test(s) => Ok(Statement::Test(self.analyze_test_statement(&s)?)),
+            ast::Statement::Expr(s) => Ok(Statement::Expr(self.analyze_expression_statement(&s)?)),
+            ast::Statement::Return(s) => Ok(Statement::Return(self.analyze_return_statement(&s)?)),
+            // TODO(affo): we should fix this to include MemberAssignement.
+            //  The error lies in AST: the Statement enum does not include that.
+            //  This is not a problem when parsing, because we parse it only in the option assignment case,
+            //  and we return an OptionStmt, which is a Statement.
+            ast::Statement::Variable(s) => {
+                Ok(Statement::Variable(self.analyze_variable_assignment(&s)?))
+            }
+            // This should not happen, but it could for BadStatement.
+            _ => Err("unsupported statement in semantic analysis".to_string()),
```

Comment:
Similar to `analyze_expression` but the only option here is that we get a `BadStatement` node right? If that's the case, can we explicitly match on `BadStatement`?

To be clear, I think we should still error, I just prefer the stronger compile time guarantees that matching on all cases gives us.

Reply:
Yeah, wise 👍